### ISSUE TITLE
Tracking NameResolverProvider being experimental

### ISSUE THREAD

**Issue by dapengzhang0** [grpc/grpc-java#4159]

**Comment by carl-mastrangelo** [comment]
Some complaints for when this is eventually reviewed:

* NRP is a factory, but contains a method called `asFactory`.  This is confusing.  
* `asFactory()` is not final, so it isn't clear what subclasses should do.  Override it?  
* `asFactory` is only called from internal classes today, maybe it shouldn't have been exposed.
* `NameResolver.Factory`, the super class is experimental.  This forces NRP to expose the static constant `PARAMS_DEFAULT_PORT`, which is currently the same as the super class.

I am convinced that mixing non-private, static methods and variables with non-final classes is a mistake.   I wish they were separate classes.

**Comment by ejona86** [comment]
The Factory thing is a hold-over from the older API, everything related to it should be removed.

I mentioned to @zhangkun83 earlier, and he seemed to agree, that the `Attributes` passed to `newNameResolver(URI, Attributes)` is in appropriate and mainly a hold-over from when we created it. We should create a custom object instead, like we did with `LoadBalancer.PickSubchannelArgs`.

**Comment by zhangkun83** [comment]
I agree with @ejona86 that we should switch away from the `Attributes` to a custom object.

**Comment by zhangkun83** [comment]
> I agree with @ejona86 that we should switch away from the Attributes to a custom object.

@ejona86, the in-process transport won't set the "default port", and doesn't need to set the proxy detector either (although it would be set set anyway because #5173 has it done in the channel). What do you feel about having optional fields in the custom object? Is it still better than Attributes?

**Comment by ejona86** [comment]
We are the only/main implementer of the custom object. Having proxy detector and default port there doesn't seem a problem; a name resolver doesn't have to use it, but they should be able to use it if they need it (that is, it is optional to use, but not as optional to be present).

**Comment by larry-safran** [comment]
This can't be stabilized until NameResolver is stabilized.
NameResolver.Factory should eventually go away and `newNameResolver` migrated to `NameResolverProvider`.
